### PR TITLE
Split generated constant init methods when the methods get larger 

### LINKED ISSUE #38941: [Bug]: Method too large exception in a Ballerina method with large amounts of constants defined
### Description

In my Ballerina program, there is an `init()` method throwing the following error.

```
ERROR [healthcare_fhir_r4.bal:(18:1,3045:2)] method is too large: '$split$method$3'
error: compilation contains errors
```

### Steps to Reproduce

**Steps to reproduce and workarounds tried**
- The mentioned `init()` method had a constant initialization, which caused around 3000 lines of code. Given the readability concerns of the code, the constants were exported to a separate `bal` file called `variable.bal` and referred to that in `init()`  method. However, yet it caused the same issue. 
- Then again, `bal build` was tried with removing the constant reference (defined in `variable.bal`) in `init()` and for our surprise, the issue persisted. 
- Therefore, tried reducing the amount of defined constants to much less number (around 1000 lines of code) in the `variable.bal` file and observed that the **build is happening**. 

So irrespective of whether it is called/referred in a function, Ballerina throws this error when there are large set of constants defined somewhere in the code (even in a separate `.bal` file)

### Affected Version(s)

Ballerina distributions: _2201.1.2_,  _2201.2.3_

### OS, DB, other environment details and versions

Ubuntu 20.04

### Related area

-> Compilation

### Related issue(s) (optional)

#38796 

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #39019: [Bug]: Method too large error when large number of constants defined in a module
### Description

When there are several `const` values defined in a Ballerina project, Method too large error occurs (observed with about 5000 constants across `bal` files in a module).

Stack trace: 
```
[2022-12-12 14:11:37,691] SEVERE {b7a.log.crash} - Method too large: gabilan/large_class/0/constants/$_typeref_type_constants.$typeref_type_init0 ()V 
org.objectweb.asm.MethodTooLargeException: Method too large: gabilan/large_class/0/constants/$_typeref_type_constants.$typeref_type_init0 ()V
        at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2087)
        at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:447)
        at org.wso2.ballerinalang.compiler.bir.codegen.split.constants.JvmRefTypeConstantsGen.generateClass(JvmRefTypeConstantsGen.java:130)
        at org.wso2.ballerinalang.compiler.bir.codegen.split.JvmConstantsGen.generateConstants(JvmConstantsGen.java:100)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:810)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:96)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:71)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:309)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:476)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:384)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:165)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:134)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:112)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:171)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:111)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:165)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:213)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

### Steps to Reproduce

Define large number of `const` in a module and build.

### Affected Version(s)

2201.1.x, 2201.2.x, 2201.3.x

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #39019 
Fixes #38941

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java [hunk 1]
```diff
@@ -90,10 +96,19 @@ public String add(BArrayType arrayType) {
 
     private String generateBArrayInits(BArrayType arrayType) {
         String varName = JvmConstants.ARRAY_TYPE_VAR_PREFIX + constantIndex++;
+        if (arrayTypeVarCount % MAX_CONSTANTS_PER_METHOD == 0 && arrayTypeVarCount != 0) {
+            mv.visitMethodInsn(INVOKESTATIC, arrayConstantsClass,
+                    B_ARRAY_TYPE_INIT_METHOD + methodCount, "()V", false);
+            genMethodReturn(mv);
+
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java:103]: Do we need space here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java [hunk 2]
```diff
@@ -107,6 +122,26 @@ private void genPopulateMethod(BArrayType arrayType, String varName) {
...
+        int populateFuncCount = 0;
+        int populateInitMethodCount = 1;
+        MethodVisitor mv = cw.visitMethod(ACC_PUBLIC + ACC_STATIC, B_ARRAY_TYPE_POPULATE_METHOD,
+                "()V", null, null);
+        for (String funcName : funcNames) {
+            if (populateFuncCount % MAX_CONSTANTS_PER_METHOD == 0 && populateFuncCount != 0) {
+                mv.visitMethodInsn(INVOKESTATIC, arrayConstantsClass,
+                        B_ARRAY_TYPE_POPULATE_METHOD + populateInitMethodCount, "()V", false);
+                genMethodReturn(mv);
+
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java:135]: We can remove unwanted spaces

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java (hunk 1 above)
Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java:103]: We can remove in other places as well.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java [hunk 3]
```diff
@@ -73,6 +76,8 @@ public JvmArrayTypeConstantsGen(PackageID packageID, BTypeHashComparator bTypeHa
         this.arrayTypeVarMap = new TreeMap<>(bTypeHashComparator);
         this.funcNames = new ArrayList<>();
         this.types = types;
+        arrayTypeVarCount = 0;
+        methodCount = 1;
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmArrayTypeConstantsGen.java:80]: Shall we move the initialisation to the declaration? It is preferred due to [this](https://stackoverflow.com/questions/3918578/should-i-initialize-variable-within-constructor-or-outside-constructor). We can change in the other places also. Same can be done for `funcName` array list.

File: tests/jballerina-unit-test/src/test/resources/test-src/jvm/largePackage/modules/errors/distinct_errors.bal [hunk 4]
```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/resources/test-src/jvm/largePackage/modules/errors/distinct_errors.bal:1]:
```suggestion
// Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```

**Review by Nadeeshan96 — APPROVED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**